{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pubmed_parser as pp\n",
    "from bs4 import BeautifulSoup \n",
    "from urllib.request import urlopen\n",
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_xml = pp.list_xml_path('C:\\\\Users\\\\KHANG\\\\Desktop\\\\test')\n",
    "pubmed_dict = pp.parse_pubmed_xml(path_xml[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "dicts_out = pp.parse_pubmed_paragraph('C:\\\\Users\\\\KHANG\\\\Desktop\\\\test\\\\peerj-06-4322.nxml', all_paragraph=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-43',\n",
       "   'ref-58',\n",
       "   'ref-65',\n",
       "   'ref-28',\n",
       "   'ref-62',\n",
       "   'ref-51',\n",
       "   'ref-33'],\n",
       "  'section': 'Introduction',\n",
       "  'text': 'Convergent evolution provides some of the strongest support for the theory of evolution through natural selection. In the case of evolutionary convergence, organisms that may have very different evolutionary histories (as measured phylogenetically), are driven by similar selective pressures to a highly similar phenotype (Losos, 2011). These selective pressures repeatedly create the same adaptive syndrome—a set of characteristics which come together to allow a specific lifestyle or perform a certain task (Reich et al., 2003). In many instances in the past, convergent evolutionary syndromes have confounded taxonomists, who (for example) mistakenly grouped New-World and Old-World vultures (Seibold & Helbig, 1995), all marine mammals (Foote et al., 2015), and many disparate lineages of microscopic organisms (Scamardella, 1999; Palenik & Haselkorn, 1992; Gupta, 2000), into clades which ultimately proved to be paraphyletic.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-74',\n",
       "   'ref-12',\n",
       "   'ref-23',\n",
       "   'ref-7',\n",
       "   'ref-56',\n",
       "   'ref-79',\n",
       "   'ref-3',\n",
       "   'ref-36'],\n",
       "  'section': 'Introduction',\n",
       "  'text': 'While phenotypic features of convergent taxa will appear superficially similar, they are not expected to share genomic similarity due to their evolutionary independence. Large number of possible sequence combinations can result in the same protein (Storz, 2016) and potentially large number of protein forms and combinations of multiple proteins that can produce the same effect (Bork, Sander & Valencia, 1993; Doolittle, 1994), so objectively defining an evolutionary syndrome using genomic data is challenging. One possible solution is to define these syndromes as a set of discrete functions rather than as a set of nucleotide sequences. In this way, convergent syndromes are described in the same way they have evolved—adaptively by function—and can be evaluated as convergent or not based on sequence similarity. Gene Ontology (GO) coding (Ashburner et al., 2000) provides an objective system by which to achieve this goal. By designating numerical codes for all possible biological activities and components, ranked hierarchically from general to specific, synonymy of function can easily be measured between even distantly related organisms throughout this text, when a discrete GO term is being referenced, it will be presented in italics, whereas when functions are being referenced in the more general sense, it will be presented in plain text. Using either experimentally determined gene/protein function or sequence similarity to previously identified functions, the activities of individual genes are paired with specific numeric codes. Gene Ontology analyses have been used in other studies to determine the functional components to a variety of traits, adaptations and physiologies of interest, including adaptation to high altitudes (Qiu et al., 2012), depth tolerance in deep-sea bacteria (Vezzi et al., 2005), and a number of human disorders (Ahn et al., 2003; Holmans et al., 2009); however, these have identified known genes of interest and then drawn conclusions of function post hoc. Rather than assigning the Gene Ontology codes first and subsequently determining the functions of particular interest as has been done previously, we can select functions of expected relevance a priori in order to allow for quantitative testing of their adaptive relevance by comparing functions in genomes in species that exhibit a convergent function. To the best of our knowledge, this is a novel approach.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-53', 'ref-30', 'ref-42'],\n",
       "  'section': 'Plant carnivory',\n",
       "  'text': 'One particularly notable convergent polyphyletic group is that of the carnivorous (alternatively, insectivorous) plants. Carnivorous plant taxa were originally classified as a single group due to their most striking and apparent feature, while disregarding features that would typically be used to define a botanical group (e.g.,\\xa0floral morphology; Primack, 1987). Subsequent work has demonstrated that a substantial number of phylogenetically distant plant lineages have evolved a carnivorous lifestyle (Givnish, 2015), presumably in response to similar selective pressures. As different lineages (or branches of the same lineage) have approached the process of insect trapping and digestion in different ways, this has in some cases made the defining of a plant as carnivorous or non-carnivorous difficult (Lloyd, 1934).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-31',\n",
       "   'ref-30',\n",
       "   'ref-35',\n",
       "   'ref-10',\n",
       "   'ref-26',\n",
       "   'ref-27'],\n",
       "  'section': 'Plant carnivory',\n",
       "  'text': 'Givnish et al. (1984) defines a carnivorous plant as one that fulfills two requirements: it must gain some detectable fitness benefit from animal remains in contact with its surfaces, and it must possess adaptations that facilitate the attraction, capture, or digestion of these prey animals. By considering only functional attributes, this definition allows a wide range of variability in the evolutionary histories and routes of adaptation of plants that are considered carnivores. Currently, nearly 600 angiosperm species are recognized as carnivorous, representing as many as nine independent origins across five families (Givnish, 2015). In addition, investigations into possible carnivorous traits in non-vascular plants such as liverworts are ongoing (Hess, Frahm & Theisen, 2005), suggesting that evolutionary shifts in nutrient acquisition strategies are perhaps even more common that currently recognized. The multiple origins and evolutionary convergence demonstrated by radiations such as those in Nepenthes and Sarracenia indicate that plant carnivory is not phylogenetically constrained; rather, it is likely that these plants are limited by their specific nutrient economics (Bloom, Chapin & Mooney, 1985), which allow them to outcompete more typical nutrient acquisition strategies only in specific habitats (Ellison & Gotelli, 2001; Ellison et al., 2003).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-25', 'ref-24', 'ref-1', 'ref-31', 'ref-26'],\n",
       "  'section': 'Plant carnivory',\n",
       "  'text': 'Carnivorous plants occupy habitat where there is little competition for sunlight. Previous studies have shown that, by leaf mass, many carnivorous plants have poor photosynthetic yield (Ellison & Farnsworth, 2005; Ellison, 2006), a likely consequence of the adaptions of their leaves for the capture of insect prey. Additionally, some carnivorous plants invest photosynthetic carbon in the fluids or secretions utilized for prey capture. In Drosera, some 3–6% this carbon, which would otherwise be expended on reproduction or vegetative growth, is used to capture prey (Adamec, 2002). As a result of these compromises, carnivorous plants only outcompete other plants in habitat where the resources that they sacrifice as a consequence of the carnivorous lifestyle (carbon, water, sunlight) are plentiful, while the resources they specialize in obtaining (nitrogen, phosphorus) are scarce. These environments are likely to be wet and sunny, with acidic, nutrient-deficient soils (Givnish et al., 1984; Ellison & Gotelli, 2001).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-64',\n",
       "   'ref-29',\n",
       "   'ref-34',\n",
       "   'ref-59',\n",
       "   'ref-45',\n",
       "   'ref-66',\n",
       "   'ref-29',\n",
       "   'ref-39',\n",
       "   'ref-15',\n",
       "   'ref-44',\n",
       "   'ref-6',\n",
       "   'ref-31',\n",
       "   'ref-31'],\n",
       "  'section': 'Carnivory-associated functions',\n",
       "  'text': 'The most apparent trait of carnivorous plants is their ability to break down prey items using digestive enzymes. As digesting animal tissue is presumably not in the repertoire of ancestral angiosperms, a question of interest is how these enzymes have evolved. In many cases, genes for digestive enzymes are apparent modifications of genes utilized in resistance and correspond to pre-existing pathways related to herbivores and pathogens (Schulze et al., 2012; Fukushima et al., 2017) or other processes present in most plants. Examples of such enzymes include chitinases, which were modified from anti-fungal and insect herbivore deterrence enzymes (Hatano & Hamada, 2008; Renner & Specht, 2012), proteases, likely derived from those involved in bacterial resistance (Mithöfer, 2011), and lipases, which are involved in metabolizing stored energy (Seth et al., 2014). Furthermore, it appears that enzymes with similar functions have evolved convergently in taxa with independent carnivorous origins (Fukushima et al., 2017), suggesting that it may not be difficult to evolve into the carnivorous niche. However, digestive enzymes may also be obtained through symbiotic interactions with micro- (Koopman et al., 2010; Caravieri et al., 2014) or macroorganisms (Midgley & Stock, 1998; Anderson & Midgley, 2003), suggesting that it may be possible to evolve into the carnivorous niche in part by appropriating the digestive enzymes of other species. While these plants fit Givnish et al.’s (1984) definition of carnivores, these digestion-associated genes would not be identifiable in the plant itself and thus would not contribute to functional overrepresentation in genomic analyses.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-69',\n",
       "   'ref-82',\n",
       "   'ref-37',\n",
       "   'ref-63',\n",
       "   'ref-11',\n",
       "   'ref-52'],\n",
       "  'section': 'Carnivory-associated functions',\n",
       "  'text': 'In addition to modifications or resistance genes or the appropriation of enzymes produced by symbionts, evidence suggests that genes used in nutrient transport are particularly important to the carnivorous lifestyle. Plant genomes possess as many as 10 times the number of peptide transport genes compared to other eukaryotes (Stacey et al., 2002), in addition to a wide variety of transport pathways for nitrate and ammonium (Williams & Miller, 2001). In carnivorous plants, the relative number of these pathways may be even higher. For example, in a transcriptomic analysis of Utricularia gibba L., a carnivorous bladderwort with a minute genome of only 80 megabases, 77 unique sequences corresponding to nitrogen transport were identified (Ibarra-Laclette et al., 2011). Modification and specialization has also occurred in transporters for other resources. For plants with traps involving rapid movement such as Dionaea muscipula Sol. ex J.Ellis, uptake of prey nutrients may be coupled to a trap’s electrical potential (Scherzer et al., 2013). Modified pathways for osmolite uptake have been identified in D.\\xa0muscipula, which uses the HKT1-type ion channel to absorb sodium without disrupting the action potential of the trap (Böhm et al., 2016). Similar adaptations may benefit less active traps as well, as for example in Sarracenia flava L. amino acid uptake is dependent on a potassium ion gradient (Plummer & Kethley, 1964).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-20', 'ref-78', 'ref-46', 'ref-67', 'ref-2'],\n",
       "  'section': 'Carnivory-associated functions',\n",
       "  'text': 'Genomics represent a new approach to investigate the evolution of novel organismal function. While the origin of novel biological functions and their role in adaptation to new habitats and ecological niches has been an important topic in evolutionary biology since the inception of the field (Darwin & Darwin, 1888), we now know that genes may be preferentially duplicated and modified, a common route to increased complexity and the possibility of new structures (Vandenbussche et al., 2003) and pathways (Monson, 2003). In more extreme cases, a whole-genome duplication event precedes an episode of major adaptive change (Soltis et al., 2009), leaving a lineage with thousands of redundant additional genes on which evolutionary processes can act. Gene copies with adaptive value are preferentially retained, while others are silenced and eventually lost (Adams & Wendel, 2005). If this general pattern is true of the genes involved in plant carnivory, such genes should be identifiable on the basis of function and would be expected to show a signal of overrepresentation in the genome.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-4', 'ref-76', 'ref-9', 'ref-61'],\n",
       "  'section': 'Carnivory-associated functions',\n",
       "  'text': 'Gene Ontology coding is an essential tool for resolving the issue of relating functionally similar (but non-homologous) genes—by design, genes that differ substantially in ancestry but provide the same function should be assigned the same Gene Ontology code(s). These descriptors are originally assigned based on experimental studies of specific genes in model organisms, which later allows non-experimental assignment using sequence homology; however, as automated annotation must be based on the content of a reference database, known biases in these databases must be considered. For example, studies addressing multiple genes often focus on a specific gene class within a specific organism, resulting in an overemphasis of that class in that organism and its relatives; experiment-based annotations will be far more common for model organisms or those of economic interest; and, as more sequences are assigned function through extrapolation rather than experimentation, those assignments can be further propagated, progressively increasing the distance from the original experimental basis (Altenhoff et al., 2012; Thomas et al., 2012). In particular, due to these biases and methods of accurately matching samples to references, there is concern that functional divergence may be missed in cases where divergent sequences remain similar, or conversely, that erroneous function may be assigned when there is substantial divergence from the nearest-matching reference sequence. Despite this, it has been previously shown that known functionally-divergent paralogs also diverged (by 32% on average) in GO codes assigned by automation (Blanc & Wolfe, 2004) and that genes typically retain highly similar functions at amino acid identity levels as low as 40% (Sangar et al., 2007). Thus, there is reason to believe that identifying function from sequence data should be sufficiently accurate at our desired level of specificity.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-84'],\n",
       "  'section': 'Hypotheses',\n",
       "  'text': 'This study seeks to test for a functional genetic signal of evolutionary convergence at the level of the genome. Specifically, it seeks to test whether or not a convergently evolved functional syndrome (i.e.,\\xa0metabolic pathways of carnivory) will rely on the same functions across lineages (as seen in Yang et al., 2015). Three possibilities will be considered. First, organisms sharing this syndrome may not be genomically distinct from others. This is possible if the functional changes required for this syndrome are not substantial at the genome level (e.g.,\\xa0changes based on slight modification of regulatory elements or alternative splicing), or if neutral variation among taxa is so substantial that the changes fall within the range of normal lineages. In this case, no signal should be detected differentiating experimental taxa from control samples (i.e.,\\xa0GO codes matching to expected carnivory-associated functions are not overrepresented). Second, a syndrome may require a specific set of functions at high representational levels in every lineage where it arises. This would be expected if the use of certain molecular machinery were unavoidable for a task, preventing evolution of the syndrome by any other pathways. In this case, it would be expected that a strong signal would be detected for functions across all experimental taxa (i.e.,\\xa0GO codes matching to expected carnivory-associated functions are uniformly overrepresented across carnivorous taxa). Lastly, a syndrome may indeed make use of some functions from a set list each time it arises, but not necessarily the same functions in each case. This would occur where there are several ways to address the same problem. A result where many of the predicted functions show strong signal, but with greatly different findings in each taxon, would support this model (i.e.,\\xa0GO codes matching to different carnivory-associated functions are overrepresented in each carnivorous taxon).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-50',\n",
       "   'ref-5',\n",
       "   'ref-63',\n",
       "   'ref-11',\n",
       "   'ref-37',\n",
       "   'ref-64',\n",
       "   'ref-60',\n",
       "   'supp-3',\n",
       "   'ref-64',\n",
       "   'ref-50',\n",
       "   'ref-11',\n",
       "   'ref-60',\n",
       "   'ref-5',\n",
       "   'ref-63',\n",
       "   'ref-37',\n",
       "   'ref-8',\n",
       "   'table-1'],\n",
       "  'section': 'Identification of carnivory-associated functions',\n",
       "  'text': 'A literature review was conducted to develop a reference set of functions previously found to be associated with plant carnivory. A topic search was performed on Web of Science in December, 2016 with the following parameters: (“carnivorous plant” OR “insectivorous plant”) AND (“gene” OR “genome” OR “transcriptome” OR “protein”) AND (“digestion” OR “transport”), producing 21 results. Publications discussing specific genes (Owen, 1999; An, Fukusaki & Kobayashi, 2002; Scherzer et al., 2013; Böhm et al., 2016) or overviews of putatively carnivory-associated genes (Ibarra-Laclette et al., 2011; Schulze et al., 2012; Rottloff et al., 2016) in sequenced carnivorous plant taxa were included (details in Table S1). From Dionaea muscipula, functions of proteins identified via proteomic analysis of the trap fluid (Schulze et al., 2012) were listed, with the addition of genes related to transport that had been specifically targeted by other studies (Owen, 1999; Böhm et al., 2016). Similarly, in Nepenthes, proteomic analysis of trap fluid released a list of functions likely to be associated with plant carnivory (Rottloff et al., 2016) with other studies assaying for a specific digestion-associated enzyme and detecting transport activity via traps’ glandular symplasts (An, Fukusaki & Kobayashi, 2002; Scherzer et al., 2013, respectively). In the bladderwort Utricularia gibba L., transcriptomic analysis was used to detect statistically increased expression of genes in traps and leaves putatively associated with carnivory (Ibarra-Laclette et al., 2011). Gene function terms, as given by these publications, were cross-referenced with the AmiGO2 Gene Ontology Database (Balsa-Canto et al., 2016) and matched to discrete GO codes that accurately represent their functions (Table 1). Of 54 total terms selected, 36 final GO codes were matched, with five terms synonymized and combined with matched terms and 13 terms having no appropriate match.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-8', 'supp-3'],\n",
       "  'section': 'Carnivory-associated functions identified via literature review.',\n",
       "  'text': 'Functions were matched to Gene Ontology terms and codes using the AmiGO2 database (Balsa-Canto et al., 2016). In cases where multiple GO codes are given, they are equivalent to or deprecated from the best-matching current term. See Table S1 for more information.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-29', 'ref-14', 'ref-41', 'ref-40', 'fig-1'],\n",
       "  'section': 'Taxon sampling',\n",
       "  'text': 'GenBank’s list of assessed plant genomes was surveyed for the inclusion of plants historically considered to be carnivorous. Four were available: Cephalotus follicularis (Fukushima et al., 2017), Drosera capensis (Butts, Bierma & Martin, 2016), Genlisea aurea (Leushkin et al., 2013), and Utricularia gibba (Lan et al., 2017). The carnivorous taxa sampled represent three independent origins of plant carnivory (Genlisea and Utricularia likely sharing a single origin) in three plant orders (Caryophyllales, Oxalidales, and Lamiales). They also exemplify four different strategies for prey-capture. Cephalotus is a pitcher/pitfall trap, using a nectar lure, slippery rim, and downward-facing projections to guide prey into a digestive soup and prevent their escape; this strategy is also seen in Nepenthes, most Sarracenia, and some carnivorous bromeliads. Drosera is a sticky-trap plant, with glandular trichomes on its leaves that secrete both sticky compounds to prevent prey’s escape and digestive enzymes to break them down; Pinguicula and Byblis also use this strategy. Genlisea is considered a lobster-pot trap, where prey species are guided to a small, funnel-like opening, through which exit is impossible; Sarracenia psittacina and, arguably, Darlingtonia californica employ this strategy. Lastly, Utricularia gibba, an aquatic carnivorous plant, uses a number of air-filled bladders to capture and digest prey. A trigger hair is stimulated as potential prey investigates the trap, releasing an air bubble contained within; the resulting vacuum pulls the prey inside, and the trap closes behind them. While no other carnivorous taxa possess this specific form, it does share some characteristics (a fast-moving trap activated by the stimulation of a trigger hair) with Aldrovanda and Dionaea muscipula. The trap characteristics, floral morphology, and overall growth form of the carnivorous taxa included in this study are depicted in Fig. 1.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-75',\n",
       "   'ref-83',\n",
       "   'ref-38',\n",
       "   'ref-55',\n",
       "   'ref-49',\n",
       "   'ref-77',\n",
       "   'ref-75',\n",
       "   'fig-2'],\n",
       "  'section': 'Taxon sampling',\n",
       "  'text': 'Non-carnivorous plants were also surveyed in order to establish a control range of “typical” flowering plants. All assessed plant genomes for which Gene Ontology-coded annotations are already available were included: Arabidopsis thaliana (Swarbreck et al., 2008), Boea hygrometrica (Xiao et al., 2015), Glycine soja (Kim et al., 2010; Qi et al., 2014), and Oryza sativa (Ohyanagi, 2006). Note that one of the carnivorous taxa, Genlisea aurea, also possessed GO annotations. Lastly, the genomes of the two non-carnivorous plants most closely related to carnivorous taxa were included: Ocimum tenuiflorum (Upadhyay et al., 2015), closest sequenced relative of Byblis, Genlisea, Pinguicula, and Utricularia; and Actinidia chinensis (Swarbreck et al., 2008), closest sequenced relative of Darlingtonia, Heliamphora, Roridula, and Sarracenia. Boea hygrometrica, included for its available annotations, is also within the order of Genlisea and Utricularia. The reference-range taxa selected cover five orders (Brassicales, Ericales, Fabales, Lamiales, and Poales), including both Monocots and Eudicots; thus, these samples can be considered a reasonable representation of the diversity and variation of angiosperms as a whole (Fig. 2). While pairwise sampling and analysis of related carnivorous and non-carnivorous taxa would be optimal to explicitly control for phylogenetic effects, this is unfortunately not possible at present due to the lack of sequenced genomes for many plant orders and the scarcity of annotated plant genomes in general. However, we expect our current reference sampling design, which includes both non-carnivorous representatives from several carnivore-containing orders and a wide phylogenetic range of taxa overall, to somewhat mitigate this potential source of bias.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-30', 'ref-68', 'ref-57'],\n",
       "  'section': 'Radial phylogeny of all angiosperms, indicating the position of taxa relevant to this study.',\n",
       "  'text': 'White-filled triangles indicate monophyletic plant orders. Each bar indicates one genus. Blue indicates a typical/non-carnivorous control taxon included in this study; green indicates a carnivorous taxon included in this study; orange indicates a carnivorous genus (as listed in Givnish, 2015) for which no genome is available. Created from tree data found in Soltis et al. (2011), visualized in FigTree (Rambaut, 2009) and manually edited in InkScape.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-81', 'ref-48', 'ref-54', 'ref-18', 'fig-3'],\n",
       "  'section': 'Data processing',\n",
       "  'text': 'For taxa lacking GO annotation but having putative genes already identified (e.g., Cephalotus follicularis), FASTA-formatted amino acid sequence data was downloaded. The remaining samples (Actinidia chinensis, Ocimum tenuiflorum, Drosera capensis, and Utricularia gibba) lacked any usable annotation data. While ideally genes and gene functions are predicted by in-depth transcriptomic studies, the training of species-specific gene identification models, and then confirmed by individual-gene experimental studies, this is simply unfeasible for studies of diverse sets of non-model taxa. Instead, predictions of genes had to be made on the simpler basis of reading frame detection. Unannotated genomes were downloaded as FASTA-formatted nucleotide sequence and processed with ORFFinder (Wheeler et al., 2003) using parameters: (-ml 450 –n false). These parameters were selected to identify putative genes and extract the predicted amino acid sequence. While some error in gene prediction are still likely from this method, parameters were set with the hope of preventing truncated or erroneously-predicted genes from entering the pipeline, e.g., very short of less than 150 amino acids and those contained entirely within the reading frame of another longer gene. Amino acid sequence data was then analyzed via BLAST-P on the Ohio Supercomputer (Ohio Supercomputer Center, 1987) with the following parameters: (-db nr -task blastp-fast -seg yes -num_alignments 10 -max_hsps 2 -evalue 1e-3), searching against the non-redundant protein sequence database (Pruitt, Tatusova & Maglott, 2007). BLAST outputs were imported into Blast2GO (Conesa et al., 2005) and matched to GO codes using the automated “Mapping” function. Exported mapping results were then processed via the custom “AnnotationConverter.pl” script, to convert data into a more accessible simplified text format. For taxa already accompanied by GO-coded gene annotations (Arabidopsis thaliana, Boea hygrometrica, Glycine soja, Oryza sativa, and Genlisea aurea), GenBank GBFF files were downloaded. The custom Perl script “GBFFConverter.pl” was used to extract genes with associated GO information as simplified text. Using the “Functionalizer.pl” Perl script, the resulting text data was then scanned for GO codes matching to the hypothesized carnivory-associated functions selected. Counts of carnivory-associated genes were weighted against total number of genes for which at least one function could be assigned, with the resulting proportions (count of function, per thousand genes) used for subsequent statistical analyses. This process is summarized graphically in Fig. 3. Putative genes that could not be assigned to any function, or that were assigned functions that could not be mapped to any GO codes, were not included in total gene counts or proportional weighting of data. By using a conservative E-value parameter in BLAST assignment of protein functions, we hoped to filter out low-certainty annotations, particularly those potentially arising from erroneously-predicted protein sequences. Following this process, a data normalization step was performed to correct for differences in tendency to detect certain functions in BLAST searches vs. from GenBank annotation data.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['supp-11'],\n",
       "  'section': 'Flowchart detailing the preparation and processing steps to obtain gene function representation data used for subsequent statistical analyses.',\n",
       "  'text': 'Solid lines indicate processing of sampled taxa, while dashed lines indicate preparation of the reference functional set by which the taxa will be evaluated. Green boxes indicate stages utilizing custom data-processing scripts (available in Supplemental Information 4).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['supp-4'],\n",
       "  'section': 'Data processing',\n",
       "  'text': 'To correct for differences in the likelihood of assigning a given function between samples accompanied by previous annotation and those coded using BLAST and mapping, Arabidopsis thaliana was analyzed by both methods, with the additional data set following the nucleotide sequence data preparation steps detailed above. The raw results of BLAST-annotated data were then multiplied by the quotient of the pre-annotated data results and the A. thaliana BLAST data results to produce corrected gene representation data. These data, along with pre-annotated samples that did not require correction, were used in all statistical tests; raw data were subjected to the same analyses, to ensure that the magnitude of changes in results would not be extreme (suggesting the need for more complex methods of error correction). The overall assessment of carnivory-associated function (“Total Carnivorous” vs. “None of the Above”) was recalculated for each taxon from the adjusted values of each function and the total gene count (“Total”). Statistical significance was considered to have six levels (“NS”, “.”, “*”, “**”, “***”, “****”); number of levels changed—either increasing or decreasing—were noted. The raw values used in these corrections are listed in Table S2.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-72', 'ref-73', 'ref-19'],\n",
       "  'section': 'Statistical analyses',\n",
       "  'text': 'Species were divided into “carnivorous” and “non-carnivorous” groups and analyzed on 25 criteria (24 carnivory-associated functions, plus the sum representation of all carnivory-associated functions in the genome) using a series of upper-tailed t-tests. To correct for multiple tests, Storey’s correction, which uses a Bayesian approach to determine realistic false discovery rate (FDR) for the numerous tests involved in genome-wide studies (Storey, 2003; Storey & Tibshirani, 2003; Dabney, Storey & Warnes, 2010) was applied, with resulting q-values used to assess significance (α\\xa0=\\xa00.05).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['table-2'],\n",
       "  'section': 'Genome information',\n",
       "  'text': 'A general assessment of all included genomes for genome size, total gene number, and number of unique Gene Ontology codes identified (a representation of diversity of functions encoded) showed largely overlapping ranges of values (Table 2). Both the largest and smallest genomes analyzed were to carnivorous plants: Genlisea aurea at 43.3 Mb and Cephalotus follicularis with 1.6 Gb (non-carnivorous plants ranged from 119.7 Mb to 1.5 Gb). Similar results were found for number of genes encoded, ranging from 17,685 in G. aurea to 89,073 in Drosera capensis (typical plants: 28,382 to 70,250). The largest number of unique GO codes identified was found in Arabidopsis thaliana; however, this results from the utilization of A. thaliana in the development of plant GO codes. The smallest number was found in Oryza sativa (1262), with the largest (after A. thaliana) found in C. follicularis (4664). Interestingly, when testing for relationships between these factors, no significant (α\\xa0=\\xa00.05 associations were found between genome size and gene number (p\\xa0=\\xa00.857, R2\\xa0=\\xa00.1202), genome size and number of unique GOs (p\\xa0=\\xa00.630 R2\\xa0=\\xa00.0909), or gene number and number of unique GOs (p\\xa0=\\xa00.901, R2\\xa0=\\xa00.1227).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['fig-4', 'supp-5'],\n",
       "  'section': 'Genome information',\n",
       "  'text': 'When comparing proportion of functionally identifiable genes that could be mapped to a carnivory associated function, there was little difference between carnivorous and non-carnivorous taxa. The percentage of genes mapping to a carnivory associated function in carnivorous taxa ranged from 3.1% to 5.0% of all function-assigned genes; in typical plants, this value ranged from 2.4% to 4.3%. In terms of which specific carnivory associated functions made up each plant’s proportion, the representation of each function varied wildly from taxon to taxon (Fig. 4; Table S3).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['supp-4'],\n",
       "  'section': 'Chart of proportional representation of carnivorous functions vs. overall gene functions in all taxa sampled.',\n",
       "  'text': 'Pie-charts above indicate total proportion of all carnivorous functions combined (red & percentage) vs. all other genes for which at least one function could be identified (blue). Stacked bars below indicate the proportion ascribed to each carnivory-associated function within the total, sorted from (on average) most represented (bottom) to least represented (top). The final bar, “Other”, combines the rarest nine functions, which each on average represent only 0.7% of the carnivory-associated functions detected. A complete numerical view of this data is available in Table S2.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['table-3'],\n",
       "  'section': 'Genome information',\n",
       "  'text': 'Statistical comparisons of the genomic representation of each carnivory associated function in carnivorous vs. typical plants yielded two significant (α\\xa0=\\xa00.05 results: “Alternative oxidase activity” (t\\xa0=\\xa03.14, p\\xa0=\\xa00.011, q\\xa0=\\xa00.047) and “ATP:ADP antiporter activity” (t\\xa0=\\xa04.00, p\\xa0=\\xa04.30E−03, q\\xa0=\\xa00.037). A third function, “phospholipase activity” (t\\xa0=\\xa02.79, p\\xa0=\\xa00.019, q\\xa0=\\xa00.053), was detected as significant before correction, but retained only marginal significance (α\\xa0=\\xa00.10 after accounting for multiple tests. Detailed results of these tests are presented in Table 3. For each test, statistical power reaches 50% (β\\xa0=\\xa00.50 at an effect size of 1.06 standard deviations and 95% (β\\xa0=\\xa00.05 at an effect size of 2.12 standard deviations (α\\xa0=\\xa00.05).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['table-4'],\n",
       "  'section': 'Genome information',\n",
       "  'text': 'Testing of individual carnivorous taxa yielded a total of 13 significant (α\\xa0=\\xa00.05 results and an additional five marginal (α\\xa0=\\xa00.10 results, out of 100 total tests (4 species ×\\xa025 distributions). Genlisea aurea and Drosera capensis had very few functions that showed a signal of genomic overrepresentation. In Genlisea aurea, only a single function, “phospholipase activity” (Z\\xa0=\\xa02.76, p\\xa0=\\xa02.89E−03, q\\xa0=\\xa00.054) result reached marginal significance. Drosera capensis had one significant function: “alternative oxidase activity” (Z\\xa0=\\xa03.72, p\\xa0=\\xa01.01E−04, q\\xa0=\\xa02.45E−03). Utricularia gibba and Cephalotus follicularis were found to have a substantial portion of carnivory-associated functions showing strong signals of genomic overrepresentation. In U. gibba, five functions reached statistical significance: “alternative oxidase activity” (Z\\xa0=\\xa02.36, p\\xa0=\\xa09.17E\\xa0−\\xa003, q\\xa0=\\xa00.025), “ammonium transmembrane transport” (Z\\xa0=\\xa03.58, p\\xa0=\\xa01.74E\\xa0−\\xa004, q\\xa0=\\xa02.09E\\xa0−\\xa003), “ATPase activity” (Z\\xa0=\\xa03.19, p\\xa0=\\xa07.19E\\xa0−\\xa004, q\\xa0=\\xa04.31E\\xa0−\\xa003), “cysteine-type peptidase activity” (Z\\xa0=\\xa02.00, p\\xa0=\\xa00.023, q\\xa0=\\xa00.046), “phosphatase activity” (Z\\xa0=\\xa02.65, p\\xa0=\\xa04.05E\\xa0−\\xa003, q\\xa0=\\xa00.016), and “phospholipase activity” (Z\\xa0=\\xa02.30, p\\xa0=\\xa00.011, q\\xa0=\\xa00.025). An additional three test results were marginally significant: “aspartic-type peptidase activity” (Z\\xa0=\\xa01.75, p\\xa0=\\xa00.040, q\\xa0=\\xa00.060), “ATP:ADP antiporter activity” (Z\\xa0=\\xa01.54, p\\xa0=\\xa00.062, q\\xa0=\\xa00.083), and total proportion of carnivory-associated functions (Z\\xa0=\\xa01.79, p\\xa0=\\xa00.036, q\\xa0=\\xa00.060). In C. follicularis, seven functions reached significance: “alternative oxidase activity” (Z\\xa0=\\xa02.36, p\\xa0=\\xa09.10E\\xa0−\\xa003, q\\xa0=\\xa00.030), “beta-galactosidase activity” (Z\\xa0=\\xa04.19, p\\xa0=\\xa01.40E\\xa0−\\xa005, q\\xa0=\\xa01.99E\\xa0−\\xa004), “glutathione transferase activity” (Z\\xa0=\\xa02.22, p\\xa0=\\xa00.13, q\\xa0=\\xa00.31), “lipase activity” (Z\\xa0=\\xa02.14, p\\xa0=\\xa00.016, q\\xa0=\\xa00.033), “lipid transport” (Z\\xa0=\\xa02.31, p\\xa0=\\xa00.010, q\\xa0=\\xa00.030), “phospholipase activity” (Z\\xa0=\\xa03.50, p\\xa0=\\xa02.28E\\xa0−\\xa004, q\\xa0=\\xa00.002), and “water channel activity” (Z\\xa0=\\xa03.08, p\\xa0=\\xa01.04E\\xa0−\\xa003, q\\xa0=\\xa04.96E\\xa0−\\xa003). One additional function, “ATP:ADP antiporter activity” (Z\\xa0=\\xa01.91, p\\xa0=\\xa00.028, q\\xa0=\\xa00.050), was marginally significant. Detailed results of these tests are presented in Table 4. For each test, statistical power reaches 50% (β\\xa0=\\xa00.50 at an effect size of 2.62 standard deviations and 95% (β\\xa0=\\xa00.05 at an effect size of 4.94 standard deviations (α\\xa0=\\xa00.05).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['supp-6', 'supp-7'],\n",
       "  'section': 'Genome information',\n",
       "  'text': 'Some changes in results of the above analyses were observed when testing with the raw data sets. In analysis 1 (Table S4), a t-test comparison of carnivorous vs. non-carnivorous taxa as groups, beta-galactosidase activity, phosphatase activity, protein homodimerization, thioglucosidase activity, and water channel activity were noted as marginally significant when using uncorrected data; these values dropped below marginal significance (q <0.10) when using corrected data. Phospholipase activity was identified as significant (q\\xa0<\\xa00.05) from uncorrected data, but decreased to marginal significance after correction. In analysis 2 (Table S5), changes were as follows: Genlisea aurea: no changes. Drosera capensis: ATP:ADP antiporter activity, phosopholipase activity, and thioglucosidase activity declined from significant (q\\xa0<\\xa00.05) to NS (q\\xa0>\\xa00.10), while cysteine-type peptidase activity declined from marginal to NS. Utricularia gibba: ammonium transmembrane transport activity and ATPase activity increased from NS to ** (q\\xa0<\\xa00.01), cysteine-type peptidase activity, and phospholipase activity increased from NS to significant, and aspartic-type peptidase activity and total carnivorous function increased from NS to marginal; however, phosphatase activity declined from ** to significant, and cinnamyl-alcohol dehydrogenase activity, polygalactosidase activity, and protein homodimerization activity declined from significant to NS. Cephalotus follicularis: beta-galactosidase activity increased from NS to *** (q\\xa0<\\xa00.001), phospholipase activity and water channel activity increased from NS to **, glutathione transferase activity increase from marginal to significant, and lipase activity and lipid transferase activity increased from NS to significant; heat shock protein activity decreased from **** (q\\xa0<\\xa00.0001) to NS, ATP:ADP antiporter activity declined from ** to marginal, ATPase activity and protein homodimerization activity decreased from ** to NS, polygalactosidase activity declined from significant to NS, and fructose bisphosphate aldolase activity declined from marginal to NS.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['table-5'],\n",
       "  'section': 'Genome information',\n",
       "  'text': 'In total across all tests, 81% of results remained unchanged in designation (Table 5). For those that did change, a decrease in significance was more common (10%) than an increase (8%). Less than half of all changes (7% of total tests) were of more than a single significance level. Thus, the analyses presented in the main text are performed using the adjusted data.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['supp-6', 'table-3', 'supp-7', 'table-4'],\n",
       "  'section': 'Effects of data adjustment on statistical significance detected in results.',\n",
       "  'text': 'Change was measured in significance levels, considering six levels: q\\xa0>\\xa00.10 (NS), q\\xa0<\\xa00.10 (.), q\\xa0<\\xa00.05 (*), q\\xa0<\\xa00.01 (**), q\\xa0<\\xa00.001 (***), q\\xa0<\\xa00.0001 (****). “Increase” shows cases where a result went up one or more significance levels; “Decrease” shows cases where a result went down one or more significance levels; “Change > 1” shows cases where a result went either up or down two or more significance levels. “Class” indicates the results based off the categorical correction seen in Table S4 vs. main text Table 3. “Individual” indicates the total of individual comparison results as shown by Table S5 vs. main text Table 4; species names show these comparisons for each species considered separately. “Overall” shows the total of all tests.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-21', 'ref-17', 'ref-80', 'ref-32', 'ref-16'],\n",
       "  'section': 'Overall effects',\n",
       "  'text': 'This study sought to detect a signal of genomic overrepresentation of functions researchers had previously determined were associated with carnivory in plants. The two functions consistently identified as significantly overrepresented were “alternative oxidase activity” and “ATP:ADP anti-porter activity”. Alternative oxidase functions primarily in the mitochondria, as part of the electron transport chain. It is believed to function as a “protective” enzyme, to prevent over-oxidation in the mitochondria and can be activated in response to oxidative stress (Day & Wiskich, 1995). For this function to be of common importance to carnivorous taxa, there are three possible explanations: (i) carnivorous plants, due to their digestive function, produce larger amounts of reactive oxygen species (consistent with Chia et al., 2004), requiring more alternative oxidase to counteract their negative effects; (ii) alternative oxidase is encoded in a modified form, having been co-opted to perform a different function outside the mitochondrion; or (iii) alternative oxidase is functioning as it would typically, but due to the similar habitat parameters of carnivorous plants, they require its effects more frequently. ATP:ADP anti-porter activity has two functions: It is involved in the maintenance of cellular electrical potential (due to an H+ gradient) in the presence of free fatty acids (Vianello, Petrussa & Macrì, 1994) and it allows exchange of cellular ATP for plastid or bacterial-symbiont ADP (Greub & Raoult, 2003). In the first case, it may be responsible for interacting with the cellular proton gradient if pH changes substantially during digestion; in the second, it may provide aid to the symbiotic bacteria that assist carnivorous plants in digestion. A third function, “phospholipase activity”, was marginally significant. Phospholipases are involved in signaling interactions as well as in metabolism of fatty acids and in degrading cell membranes (Chapman, 1998). Carnivorous plants may possess an increased need for complex signaling pathways to regulate their digestive machinery, as well as a clear need to break down cell membranes to access the contents of insect cells.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-64', 'ref-60', 'ref-22', 'table-2'],\n",
       "  'section': 'Individual taxa',\n",
       "  'text': 'Utricularia gibba uniquely possessed overrepresentation in ATPase activity, cysteine-type peptidase activity, ammonium transmembrane transport, phosphatase activity, and aspartic-type endopeptidase activity. Phosphatase, aspartic peptidase, and cysteine peptidase, as catabolic enzymes found localized to the digestive fluids of other carnivorous taxa (Schulze et al., 2012; Rottloff et al., 2016), most likely have roles in direct digestive function. Ammonium transmembrane transport, while required in some amount by all plants, may be more vital for Utricularia, which must extract the concentrated nitrogenous products of digestion from an aquatic environment. ATPase in plants is involved in regulation of endocytotic and secretory processes (Dettmer et al., 2006), which would logically be involved in both the release of digestive enzymes and the absorption of digested material. Utricularia gibba was the only taxon studied that had even marginal significance in the total genomic proportion of carnivorous functions. Also of note is the vast difference in portion of carnivory-associated functions between U. gibba and its close relative Genlisea aurea. While both taxa have characteristically-reduced genomes, G. aurea has approximately half the genome size and gene number of U. gibba (Table 2). It may be that in Genlisea, selective pressure strongly favored deletion of duplicated genes, with up-regulation or modification instead occurring at the transcriptional or translational stage.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-64', 'ref-60'],\n",
       "  'section': 'Individual taxa',\n",
       "  'text': 'In Cephalotus follicularis, beta-galactosidase activity, water channel activity, glutathione transferase activity, lipase activity, and lipid transferase activity were found to be uniquely overrepresented. Lipase and beta-galactosidase (which breaks polysaccharide bonds) are likely to have direct involvement in digestion, having also been found in the digestive fluids of other carnivorous taxa (Schulze et al., 2012; Rottloff et al., 2016); lipid transferase would logically accompany lipase, either to localize lipid substrates or to move the products of their decomposition. As C. follicularis must transfer water to the interior of its pitchers for digestive functions to be possible at all, high levels of water channel activity is also a logical\\xa0finding.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-70', 'ref-71', 'ref-13', 'ref-47'],\n",
       "  'section': 'Conclusions',\n",
       "  'text': 'The degree of molecular specificity required to meet an organism’s needs can also be expected to play a role, with ability to predict a specific functional set increasing proportional to specificity of the convergent syndrome. In carnivorous plants, a wide range of morphologies (as evidenced by the taxa included in this study) have arisen to reach the same end. In other cases, there is little flexibility in how an organism can reach the needed outcome. For example, organisms that rely on the mimicry of pheromones, such as orchids that imitate bee sex and alarm pheromones, are far less likely to show variation in the functions required for the end-result (Stökl et al., 2005; Stökl et al., 2007; Brodmann et al., 2009). Conversely, even broadly-defined, frequently re-derived evolutionary syndromes may still show repeated selection for specific functional codes. It has been shown that organisms experience substantial convergence of microbiome even for classes as broad as “carnivore” vs. “herbivore” (Muegge et al., 2011); it is reasonable to consider that this occurrence may be accompanied by host genome functional convergence as well. However, to detect a signal in these broader groups, where it may be difficult to assemble a manageable list of target syndrome-associated functions, much more thorough sampling would likely be required.'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['ref-60',\n",
       "   'ref-64',\n",
       "   'ref-63',\n",
       "   'ref-50',\n",
       "   'ref-11',\n",
       "   'ref-37',\n",
       "   'ref-5'],\n",
       "  'section': 'List of protein functions identified in past studies of carnivorous plants, as published, and the taxon in which they were identified',\n",
       "  'text': 'Where zones of the plant are given (e.g., trap, fluid, or secretion), analysis was localized to that specific zone or accounted for differential expression. Where only the taxon is specified, analysis considered the whole plant. Information from: Rottloff et al. (2016) [1], Schulze et al. (2012) [2], Scherzer et al. (2013) [3], Owen (1999) [4], Böhm et al. (2016) [5], Ibarra-Laclette et al. (2011) [6], and An, Fukusaki & Kobayashi (2002) [7].'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['table-3'],\n",
       "  'section': 'Results of statistical analyses comparing non-carnivorous plants to carnivorous plants for each of 24 carnivory-associated functions, plus the total of all functions',\n",
       "  'text': 'Equivalent to main text Table 3, but using unadjusted data. “t” indicates the test statistic of an upper-tailed Student’s t-test. “p” indicates the p-value of this test. “ q” indicates a corrected p-value accounting for multiple comparisons, using Storey’s correction. Significance (“Sig.”) is indicated by bolding and with “*” for q\\xa0<\\xa00.05, “**” for q\\xa0<\\xa00.01, and “*** for q\\xa0<\\xa00.001. A non-bolded “.” indicates marginal values (q\\xa0<\\xa00.10), while “NS” indicates non-significance (q\\xa0>\\xa00.10).'},\n",
       " {'pmc': '5797450',\n",
       "  'pmid': '29404217',\n",
       "  'reference_ids': ['table-4'],\n",
       "  'section': 'Results of statistical analyses comparing non-carnivorous plants to carnivorous plants for each of 24 carnivory-associated functions, plus the total of all functions',\n",
       "  'text': 'Equivalent to main text Table 4 but using unadjusted data. “t” indicates the test statistic of an upper-tailed Student’s t-test. “ p” indicates the p-value of this test. “ q” indicates a corrected p-value accounting for multiple comparisons, using Storey’s correction. Significance (“Sig.”) is indicated by bolding and with “*” for q\\xa0<\\xa00.05, “**” for q\\xa0<\\xa00.01, and “*** for q\\xa0<\\xa00.001. A non-bolded “.” indicates marginal values (q\\xa0<\\xa00.10), while “NS” indicates non-significance (q\\xa0>\\xa00.10).'}]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dicts_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'abstract': ' Abstract   Gastrodia  kachinensis , a new species of Orchidaceae , is described and illustrated from Putao, Kachin State, Myanmar. It is morphologically similar to G.  gracilis , presumably its nearest relative, but can be easily distinguished from the latter by having perianth tube with punctate outer surface, verrucose outer surface of sepal lobe, orbicular petals, ovate-elliptic lip with truncate apex and auriculate-clawed base, glabrous lip apex with a pair of twin protuberance-like lamellae and column with a pair of blade-like lateral wings and acute stelidia at apex. Identification key and colour photographs are provided. A preliminary risk-of-extinction assessment, according to the IUCN Red List Categories and Criteria, is given for the new species. ',\n",
       " 'affiliation_list': [['A1',\n",
       "   'State Key Laboratory of Systematic and Evolutionary Botany, Institute of Botany, Chinese Academy of Sciences, Beijing 100093, China'],\n",
       "  ['A2',\n",
       "   'Southeast Asia Biodiversity Research Institute, Chinese Academy of Sciences, Yezin, Nay Pyi Taw 05282, Myanmar']],\n",
       " 'author_list': [['Aung', 'Ye Lwin', 'A1'],\n",
       "  ['Jin', 'Xiaohua', 'A1'],\n",
       "  ['Jin', 'Xiaohua', 'A2']],\n",
       " 'doi': '10.3897/phytokeys.94.21348',\n",
       " 'full_title': 'Gastrodia   kachinensis  ( Orchidaceae ), a new species from Myanmar',\n",
       " 'journal': 'PhytoKeys',\n",
       " 'pmc': '5799787',\n",
       " 'pmid': '',\n",
       " 'publication_year': '2018',\n",
       " 'publisher_id': '',\n",
       " 'subjects': 'Research Article'}"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=herb+drug+interaction&retmax=100'\n",
    "\n",
    "r = requests.get(url)\n",
    "\n",
    "encodedText = r.text.encode(\"utf-8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "soup = BeautifulSoup(encodedText,'html.parser')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29414119\n",
      "29413979\n",
      "29412757\n",
      "29412669\n",
      "29411879\n",
      "29410309\n",
      "29410154\n",
      "29403999\n",
      "29403922\n",
      "29403905\n",
      "29403896\n",
      "29403797\n",
      "29403691\n",
      "29403690\n",
      "29403686\n",
      "29409027\n",
      "29408729\n",
      "29407771\n",
      "29405809\n",
      "29405040\n",
      "29404283\n",
      "29403626\n",
      "29403307\n",
      "29399790\n",
      "29399291\n",
      "29396445\n",
      "29396335\n",
      "29393456\n",
      "29393417\n",
      "29391873\n",
      "29391436\n",
      "29389555\n",
      "29389474\n",
      "29388183\n",
      "29387831\n",
      "29387552\n",
      "29387129\n",
      "29386484\n",
      "29386163\n",
      "29385884\n",
      "29384230\n",
      "29383171\n",
      "29382035\n",
      "29382023\n",
      "29381315\n",
      "29379436\n",
      "29377739\n",
      "29376238\n",
      "29375382\n",
      "29374248\n",
      "29373834\n",
      "29373566\n",
      "29373547\n",
      "29372518\n",
      "29368030\n",
      "29367772\n",
      "29367475\n",
      "29367110\n",
      "29363722\n",
      "29363155\n",
      "29361877\n",
      "29360858\n",
      "29360539\n",
      "29360465\n",
      "29357857\n",
      "29356020\n",
      "29350757\n",
      "29350053\n",
      "29349608\n",
      "29348843\n",
      "29348768\n",
      "29348107\n",
      "29346285\n",
      "29345255\n",
      "29344421\n",
      "29344420\n",
      "29344417\n",
      "29344416\n",
      "29344413\n",
      "29343942\n",
      "29343259\n",
      "29342961\n",
      "29338725\n",
      "29337353\n",
      "29335861\n",
      "29335190\n",
      "29334999\n",
      "29334801\n",
      "29334663\n",
      "29334585\n",
      "29334256\n",
      "29333177\n",
      "29333043\n",
      "29330375\n",
      "29329613\n",
      "29328398\n",
      "29327455\n",
      "29327126\n",
      "29323793\n",
      "29322864\n"
     ]
    }
   ],
   "source": [
    "for each in soup.findAll('id') :\n",
    "    print(each.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = urlopen('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id=19008416,18787170&rettype=abstract&retmode=text')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['\\n1. Science. 2008 Dec 12;322(5908):1695-9. doi: 10.1126/science.1165395. Epub 2008\\nNov 13.',\n",
       " 'Genomic loss of microRNA-101 leads to overexpression of histone methyltransferase\\nEZH2 in cancer.',\n",
       " 'Varambally S(1), Cao Q, Mani RS, Shankar S, Wang X, Ateeq B, Laxman B, Cao X,\\nJing X, Ramnarayanan K, Brenner JC, Yu J, Kim JH, Han B, Tan P, Kumar-Sinha C,\\nLonigro RJ, Palanisamy N, Maher CA, Chinnaiyan AM.',\n",
       " 'Author information: \\n(1)Michigan Center for Translational Pathology, University of Michigan Medical\\nSchool, Ann Arbor, MI 48109, USA.',\n",
       " 'Enhancer of zeste homolog 2 (EZH2) is a mammalian histone methyltransferase that \\ncontributes to the epigenetic silencing of target genes and regulates the\\nsurvival and metastasis of cancer cells. EZH2 is overexpressed in aggressive\\nsolid tumors by mechanisms that remain unclear. Here we show that the expression \\nand function of EZH2 in cancer cell lines are inhibited by microRNA-101\\n(miR-101). Analysis of human prostate tumors revealed that miR-101 expression\\ndecreases during cancer progression, paralleling an increase in EZH2 expression. \\nOne or both of the two genomic loci encoding miR-101 were somatically lost in\\n37.5% of clinically localized prostate cancer cells (6 of 16) and 66.7% of\\nmetastatic disease cells (22 of 33). We propose that the genomic loss of miR-101 \\nin cancer leads to overexpression of EZH2 and concomitant dysregulation of\\nepigenetic pathways, resulting in cancer progression.',\n",
       " 'DOI: 10.1126/science.1165395 \\nPMCID: PMC2684823\\nPMID: 19008416  [Indexed for MEDLINE]',\n",
       " '\\n2. Science. 2008 Sep 12;321(5895):1499-502. doi: 10.1126/science.1162981.',\n",
       " 'FBXW7 targets mTOR for degradation and cooperates with PTEN in tumor suppression.',\n",
       " 'Mao JH(1), Kim IJ, Wu D, Climent J, Kang HC, DelRosario R, Balmain A.',\n",
       " 'Author information: \\n(1)Cancer Research Institute, University of California at San Francisco, 2340\\nSutter Street, San Francisco, CA 94143, USA.',\n",
       " 'The enzyme mTOR (mammalian target of rapamycin) is a major target for therapeutic\\nintervention to treat many human diseases, including cancer, but very little is\\nknown about the processes that control levels of mTOR protein. Here, we show that\\nmTOR is targeted for ubiquitination and consequent degradation by binding to the \\ntumor suppressor protein FBXW7. Human breast cancer cell lines and primary tumors\\nshowed a reciprocal relation between loss of FBXW7 and deletion or mutation of\\nPTEN (phosphatase and tensin homolog), which also activates mTOR. Tumor cell\\nlines harboring deletions or mutations in FBXW7 are particularly sensitive to\\nrapamycin treatment, which suggests that loss of FBXW7 may be a biomarker for\\nhuman cancers susceptible to treatment with inhibitors of the mTOR pathway.',\n",
       " 'DOI: 10.1126/science.1162981 \\nPMCID: PMC2849753\\nPMID: 18787170  [Indexed for MEDLINE]',\n",
       " '']"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page = BeautifulSoup(a,'html.parser')\n",
    "page.text.split('\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['\\n1. Science. 2008 Dec 12;322(5908):1695-9. doi: 10.1126/science.1165395. Epub 2008\\nNov 13.\\n\\nGenomic loss of microRNA-101 leads to overexpression of histone methyltransferase\\nEZH2 in cancer.\\n\\nVarambally S(1), Cao Q, Mani RS, Shankar S, Wang X, Ateeq B, Laxman B, Cao X,\\nJing X, Ramnarayanan K, Brenner JC, Yu J, Kim JH, Han B, Tan P, Kumar-Sinha C,\\nLonigro RJ, Palanisamy N, Maher CA, Chinnaiyan AM.\\n\\nAuthor information: \\n(1)Michigan Center for Translational Pathology, University of Michigan Medical\\nSchool, Ann Arbor, MI 48109, USA.\\n\\nEnhancer of zeste homolog 2 (EZH2) is a mammalian histone methyltransferase that \\ncontributes to the epigenetic silencing of target genes and regulates the\\nsurvival and metastasis of cancer cells. EZH2 is overexpressed in aggressive\\nsolid tumors by mechanisms that remain unclear. Here we show that the expression \\nand function of EZH2 in cancer cell lines are inhibited by microRNA-101\\n(miR-101). Analysis of human prostate tumors revealed that miR-101 expression\\ndecreases during cancer progression, paralleling an increase in EZH2 expression. \\nOne or both of the two genomic loci encoding miR-101 were somatically lost in\\n37.5% of clinically localized prostate cancer cells (6 of 16) and 66.7% of\\nmetastatic disease cells (22 of 33). We propose that the genomic loss of miR-101 \\nin cancer leads to overexpression of EZH2 and concomitant dysregulation of\\nepigenetic pathways, resulting in cancer progression.\\n\\nDOI: 10.1126/science.1165395 \\nPMCID: PMC2684823\\nPMID: 19008416  ',\n",
       " '\\n2. Science. 2008 Sep 12;321(5895):1499-502. doi: 10.1126/science.1162981.\\n\\nFBXW7 targets mTOR for degradation and cooperates with PTEN in tumor suppression.\\n\\nMao JH(1), Kim IJ, Wu D, Climent J, Kang HC, DelRosario R, Balmain A.\\n\\nAuthor information: \\n(1)Cancer Research Institute, University of California at San Francisco, 2340\\nSutter Street, San Francisco, CA 94143, USA.\\n\\nThe enzyme mTOR (mammalian target of rapamycin) is a major target for therapeutic\\nintervention to treat many human diseases, including cancer, but very little is\\nknown about the processes that control levels of mTOR protein. Here, we show that\\nmTOR is targeted for ubiquitination and consequent degradation by binding to the \\ntumor suppressor protein FBXW7. Human breast cancer cell lines and primary tumors\\nshowed a reciprocal relation between loss of FBXW7 and deletion or mutation of\\nPTEN (phosphatase and tensin homolog), which also activates mTOR. Tumor cell\\nlines harboring deletions or mutations in FBXW7 are particularly sensitive to\\nrapamycin treatment, which suggests that loss of FBXW7 may be a biomarker for\\nhuman cancers susceptible to treatment with inhibitors of the mTOR pathway.\\n\\nDOI: 10.1126/science.1162981 \\nPMCID: PMC2849753\\nPMID: 18787170  ',\n",
       " '']"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page.text.split('[Indexed for MEDLINE]\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
